File: app/controllers/plex_auth_controller.rb
class PlexAuthController < ApplicationController
  require "browser"

  PLEX_BASE_URL = 'https://plex.tv'.freeze
  PLEX_PRODUCT = 'Voterr'.freeze
  PLEX_VERSION = 'Plex OAuth'.freeze
  VOTERR_CLIENT_ID = 'voterr-8a7b6c5d-4e3f-2g1h-9i8j-7k6l5m4n3o2p'

  def new
    @plex_product = "Voterr"
    @plex_version = "1.0"
    @client_id = VOTERR_CLIENT_ID
    @browser_name = browser.name
    @browser_version = browser.version
    @device = browser.platform.name
    @device_name = "#{browser.name} (Voterr)"
    @callback_url = callback_plex_auth_url
  end

  def callback
    auth_token = params.dig(:plex_auth, :auth_token)
    
    if auth_token
      user_data = fetch_plex_user_data(auth_token)
      if user_data
        user = User.find_or_initialize_by(email: user_data[:email])
        user.update!(
          plex_token: auth_token,
          name: user_data[:username]
        )
        FetchAndStoreMoviesJob.perform_later(user)
        session[:user_id] = user.id
        render json: { success: true, message: 'Successfully authenticated with Plex!' }
      else
        render json: { success: false, message: 'Failed to fetch user data from Plex.' }, status: :unprocessable_entity
      end
    else
      render json: { success: false, message: 'No auth token provided.' }, status: :bad_request
    end
  end

  private

  def plex_headers
    {
      'Accept' => 'application/json',
      'X-Plex-Product' => PLEX_PRODUCT,
      'X-Plex-Version' => PLEX_VERSION,
      'X-Plex-Client-Identifier' => VOTERR_CLIENT_ID
    }
  end

  def fetch_plex_user_data(auth_token)
    response = plex_request(:get, '/api/v2/user', auth_token: auth_token)
    if response.status == 200
      data = JSON.parse(response.body)
      {
        email: data['email'],
        username: data['username']
      }
    end
  end

  def plex_request(method, endpoint, auth_token: nil, params: {})
    url = "#{PLEX_BASE_URL}#{endpoint}"
    response = Faraday.send(method, url) do |req|
      req.headers.merge!(plex_headers)
      req.headers['X-Plex-Token'] = auth_token if auth_token
      req.params.merge!(params)
    end

    response
  end
end

File: app/javascript/controllers/plex_auth_controller.js
import { Controller } from "@hotwired/stimulus"

export default class extends Controller {
  static targets = [ "status" ]
  
  static values = {
    plexBaseUrl: String,
    plexProduct: String,
    plexVersion: String,
    clientId: String,
    browserName: String,
    browserVersion: String,
    device: String,
    deviceName: String,
    callbackUrl: String
  }

  async initiate() {
    this.statusTarget.textContent = "Initiating Plex authentication..."
    try {
      const pin = await this.generatePlexPin()
      const authUrl = this.constructAuthUrl(pin.code)
      window.open(authUrl, '_blank')
      await this.checkPinStatus(pin.id)
    } catch (error) {
      console.error('Plex authentication failed:', error)
      this.statusTarget.textContent = 'Failed to authenticate with Plex. Please try again.'
    }
  }

  async generatePlexPin() {
    const response = await fetch(`${this.plexBaseUrlValue}/api/v2/pins`, {
      method: 'POST',
      headers: {
        'Accept': 'application/json',
        'X-Plex-Product': this.plexProductValue,
        'X-Plex-Version': this.plexVersionValue,
        'X-Plex-Client-Identifier': this.clientIdValue
      },
      body: new URLSearchParams({ 'strong': 'true' })
    })

    if (!response.ok) {
      throw new Error('Failed to generate Plex PIN')
    }

    return response.json()
  }

  constructAuthUrl(pinCode) {
    const params = new URLSearchParams({
      clientID: this.clientIdValue,
      code: pinCode,
      'context[device][product]': this.plexProductValue,
      'context[device][version]': this.plexVersionValue,
      'context[device][platform]': this.browserNameValue,
      'context[device][platformVersion]': this.browserVersionValue,
      'context[device][device]': this.deviceValue,
      'context[device][deviceName]': this.deviceNameValue,
      'context[device][model]': 'Plex OAuth'
    })

    return `https://app.plex.tv/auth#!?${params.toString()}`
  }

  async checkPinStatus(pinId) {
    this.statusTarget.textContent = "Waiting for Plex authentication..."
    const maxAttempts = 60 // 1 minute of checking
    for (let i = 0; i < maxAttempts; i++) {
      const response = await fetch(`${this.plexBaseUrlValue}/api/v2/pins/${pinId}`, {
        headers: {
          'Accept': 'application/json',
          'X-Plex-Client-Identifier': this.clientIdValue
        }
      })

      if (!response.ok) {
        throw new Error('Failed to check PIN status')
      }

      const data = await response.json()
      if (data.authToken) {
        await this.sendAuthTokenToServer(data.authToken)
        return
      }

      await new Promise(resolve => setTimeout(resolve, 1000)) // Wait 1 second before checking again
    }

    throw new Error('PIN authentication timed out')
  }

  async sendAuthTokenToServer(authToken) {
    console.log('Sending auth token to server:', authToken);
    try {
      const response = await fetch('/plex_auth/callback', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-CSRF-Token': document.querySelector('meta[name="csrf-token"]').content
        },
        body: JSON.stringify({ plex_auth: { auth_token: authToken } })
      });
  
      console.log('Response status:', response.status);
      const result = await response.json();
      console.log('Response body:', result);
      result
      if (result.success) {
        this.statusTarget.textContent = result.message;
        window.location.href = '/';
      } else {
        throw new Error(result.message);
      }
    } catch (error) {
      console.error('Error during authentication:', error);
      this.statusTarget.textContent = 'Authentication failed. Please try again.';
    }
  }
}